### PR TITLE
Add language version check to name of analyzer

### PR DESCRIPTION
**Pull request by chborl**
Fixes: #1524 

Customer scenario:
If the customer is using a version of C# or VB where nameof is not supported, the nameof analyzer will still recommend changing a string literal to `nameof()`.  If the user accepts the suggested fix, they will get an error: `CS8026: Feature 'nameof operator' is not available in C# 5.  Please use language version 6 or greater.`

Bug this fixes: 
#1524 , and added additional test support for specifying language versions

Workarounds, if any
none

Risk
None

Performance impact
None

Is this a regression from a previous update?
no

Root cause analysis
We did not consider language versions prior to C#6 or VB14 when writing the nameof analyzer

How was the bug found?
customer reported on Github

### CODE REVIEW

**Comment by chborl** [comment]
@dotnet/roslyn-analysis  and @dotnet/roslyn-ide  for review.

File: src/Microsoft.CodeQuality.Analyzers/VisualBasic/Maintainability/BasicUseNameofInPlaceOfString.vb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    ''' <summary>
+    ''' CA1507: Use nameof to express symbol names
+    ''' </summary>
+    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
+    Public NotInheritable Class BasicUseNameofInPlaceOfStringAnalyzer
+        Inherits UseNameofInPlaceOfStringAnalyzer
+
+        Friend Overrides Function ApplicableLanguageVersion(options As ParseOptions) As Boolean
+            Dim langVersion = CType(options, VisualBasicParseOptions).LanguageVersion
+            Return (langVersion > LanguageVersion.VisualBasic12)
```
Comment (jasonmalinowski) [inline src/Microsoft.CodeQuality.Analyzers/VisualBasic/Maintainability/BasicUseNameofInPlaceOfString.vb]: Parenthesis seem odd here.

File: src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/UseNameofInPlaceOfString.cs [hunk 2]
```diff
@@ -37,6 +36,8 @@ public sealed class UseNameofInPlaceOfStringAnalyzer : DiagnosticAnalyzer
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleWithSuggestion);
 
+        internal abstract bool ApplicableLanguageVersion(ParseOptions options);
```
Comment (jasonmalinowski) [inline src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/UseNameofInPlaceOfString.cs]: Why internal?

Comment (jasonmalinowski) [inline src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/UseNameofInPlaceOfString.cs]: Nit: "IsApplicableToLanguageVersion" might be a bit clearer.

File: src/Microsoft.CodeQuality.Analyzers/VisualBasic/Maintainability/BasicUseNameofInPlaceOfString.vb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+Imports Microsoft.CodeAnalysis.VisualBasic
+Imports Microsoft.CodeQuality.Analyzers.Maintainability
+
+Namespace Microsoft.CodeQuality.VisualBasic.Analyzers.Maintainability
+
+    ''' <summary>
+    ''' CA1507: Use nameof to express symbol names
+    ''' </summary>
+    <DiagnosticAnalyzer(LanguageNames.VisualBasic)>
+    Public NotInheritable Class BasicUseNameofInPlaceOfStringAnalyzer
```
Comment (jasonmalinowski) [inline src/Microsoft.CodeQuality.Analyzers/VisualBasic/Maintainability/BasicUseNameofInPlaceOfString.vb]: Dumb question: do we call these "Basic" or "VisualBasic"? Does the name show up anywhere?

File: src/Microsoft.CodeQuality.Analyzers/CSharp/Maintainability/CSharpUseNameofInPlaceOfString.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+{
+    /// <summary>
+    /// CA1507: Use nameof to express symbol names
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class CSharpUseNameofInPlaceOfStringAnalyzer : UseNameofInPlaceOfStringAnalyzer
+    {
+        internal override bool ApplicableLanguageVersion(ParseOptions options)
+        {
+            return (((CSharpParseOptions)options).LanguageVersion > LanguageVersion.CSharp5);
```
Comment (sharwell) [inline src/Microsoft.CodeQuality.Analyzers/CSharp/Maintainability/CSharpUseNameofInPlaceOfString.cs]: :bulb: consider using `>=` for this

File: src/Microsoft.CodeQuality.Analyzers/VisualBasic/Maintainability/BasicUseNameofInPlaceOfString.vb (hunk 3 above)
Comment (chborl) [reply src/Microsoft.CodeQuality.Analyzers/VisualBasic/Maintainability/BasicUseNameofInPlaceOfString.vb]: This name is consistent with the other analyzers in Roslyn-Analyzers repo.  I don't believe the name is visible anywhere, but @mavasani would know for sure.

**Pull request merged**